File: MoveTrack-iOS/MoveTrack/Services/AuthService.swift
//
//  AuthService.swift
//  Nexus Moves
//
//  Created on 2025-12-26.
//

import Foundation

class AuthService {
    static let shared = AuthService()

    private init() {}

    // MARK: - Request Magic Link
    func requestMagicLink(email: String) async throws -> MagicLinkResponse {
        struct RequestBody: Encodable {
            let email: String
        }

        let response: MagicLinkResponse = try await APIClient.shared.post(
            Constants.Endpoints.requestMagicLink,
            body: RequestBody(email: email),
            requiresAuth: false
        )

        return response
    }

    // MARK: - Verify Magic Link
    func verifyMagicLink(token: String) async throws -> AuthResponse {
        let queryItems = [URLQueryItem(name: "token", value: token)]

        let response: AuthResponse = try await APIClient.shared.get(
            Constants.Endpoints.verifyMagicLink,
            queryItems: queryItems,
            requiresAuth: false
        )

        // If successful, save session token
        if response.success, let sessionToken = response.sessionToken {
            _ = KeychainService.shared.save(sessionToken, forKey: Constants.sessionTokenKey)
        }

        return response
    }

    // MARK: - Get Current User
    func getCurrentUser() async throws -> User {
        struct MeResponse: Codable {
            let success: Bool
            let user: User?
            let error: String?
        }

        let response: MeResponse = try await APIClient.shared.get(Constants.Endpoints.me)

        guard response.success, let user = response.user else {
            throw APIError.serverError(response.error ?? "Failed to get user")
        }

        return user
    }

    // MARK: - Logout
    func logout() async throws {
        struct LogoutResponse: Codable {
            let success: Bool
            let error: String?
        }

        do {
            let _: LogoutResponse = try await APIClient.shared.post(Constants.Endpoints.logout)
        } catch {
            // Continue with logout even if API call fails
            print("Logout API call failed: \(error.localizedDescription)")
        }

        // Clear session token
        _ = KeychainService.shared.delete(forKey: Constants.sessionTokenKey)
    }

    // MARK: - Check if Logged In
    var isLoggedIn: Bool {
        return KeychainService.shared.load(forKey: Constants.sessionTokenKey) != nil
    }

    // MARK: - Get Stored Session Token
    var sessionToken: String? {
        return KeychainService.shared.load(forKey: Constants.sessionTokenKey)
    }
}

File: MoveTrack-Android/app/src/main/java/dev/we3kings/nexusmoves/data/auth/AuthService.kt
package dev.we3kings.nexusmoves.data.auth

import dev.we3kings.nexusmoves.data.Constants
import dev.we3kings.nexusmoves.data.api.ApiClient
import dev.we3kings.nexusmoves.data.api.ApiError
import dev.we3kings.nexusmoves.model.AuthResponse
import dev.we3kings.nexusmoves.model.MagicLinkResponse
import dev.we3kings.nexusmoves.model.MeResponse
import dev.we3kings.nexusmoves.model.User
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString

/**
 * Port of iOS AuthService — the OTP login flow and session lifecycle.
 * verify-code persists the returned token to [TokenStore] (Keychain parity).
 */
object AuthService {

    @Serializable private data class EmailBody(val email: String)
    @Serializable private data class VerifyBody(val email: String, val code: String)

    /** POST /auth/request-code — email the user a 6-digit login code. */
    suspend fun requestCode(email: String): MagicLinkResponse =
        ApiClient.post(
            Constants.Endpoints.requestCode,
            ApiClient.json.encodeToString(EmailBody(email)),
            requiresAuth = false,
        )

    /** POST /auth/verify-code — exchange the code for a session token. */
    suspend fun verifyCode(email: String, code: String): AuthResponse {
        val response: AuthResponse = ApiClient.post(
            Constants.Endpoints.verifyCode,
            ApiClient.json.encodeToString(VerifyBody(email, code)),
            requiresAuth = false,
        )
        if (response.success && response.sessionToken != null) {
            TokenStore.sessionToken = response.sessionToken
        }
        return response
    }

    /**
     * GET /auth/verify-magic-link?token=… — the deep-link path (a user taps a
     * magic-link email on their phone with the app installed). Persists the
     * session token on success. The app's own login is OTP; this only handles an
     * inbound magic link.
     */
    suspend fun verifyMagicLink(token: String): AuthResponse {
        val encoded = java.net.URLEncoder.encode(token, "UTF-8")
        val response: AuthResponse = ApiClient.get(
            "${Constants.Endpoints.verifyMagicLink}?token=$encoded",
            requiresAuth = false,
        )
        if (response.success && response.sessionToken != null) {
            TokenStore.sessionToken = response.sessionToken
        }
        return response
    }

    /** GET /auth/me — validate the stored token and load the user. */
    suspend fun getCurrentUser(): User {
        val response: MeResponse = ApiClient.get(Constants.Endpoints.me)
        return response.user
            ?: throw ApiError.ServerError(response.error ?: "Failed to get user")
    }

    /** POST /auth/logout — best-effort; the token is cleared regardless. */
    suspend fun logout() {
        try {
            ApiClient.post<Map<String, String>>(Constants.Endpoints.logout, "{}")
        } catch (_: Exception) {
            // Continue with logout even if the API call fails (iOS parity).
        }
        TokenStore.clear()
    }

    val isLoggedIn: Boolean get() = TokenStore.isLoggedIn
    val sessionToken: String? get() = TokenStore.sessionToken
}
